django/db/models/sql/query.py

```diff
@@ -569,6 +569,18 @@ def combine(self, rhs, connector):
         if self.distinct_fields != rhs.distinct_fields:
             raise TypeError('Cannot combine queries with different distinct fields.')
 
+        # If `lhs` and `rhs` shares the same alias prefix, it is possible to have
+        # conflicting alias changes like T4 -> T5 and T5 -> T6 (which ends up as
+        # T4 -> T6) while combining two querysets. To prevent this, change alias
+        # prefix of the `rhs` and update current aliases accordingly, except if the
+        # alias is the base table since it needs to be present in both lhs and rhs.
+        initial_alias = rhs.get_initial_alias()
+        rhs.bump_prefix(self, exclude={
+            alias for alias, join in rhs.alias_map.items()
+            if join.table_alias in [join.table_name, initial_alias]
+        })
+        rhs.unref_alias(initial_alias)
+
```

Comment:
Is it not enough to always exclude an initial alias? :thinking: New test works for me with
```diff
diff --git a/django/db/models/sql/query.py b/django/db/models/sql/query.py
index 96f34131da..23040c89ff 100644
--- a/django/db/models/sql/query.py
+++ b/django/db/models/sql/query.py
@@ -574,12 +574,7 @@ class Query(BaseExpression):
         # T4 -> T6) while combining two querysets. To prevent this, change alias
         # prefix of the `rhs` and update current aliases accordingly, except if the
         # alias is the base table since it needs to be present in both lhs and rhs.
-        initial_alias = rhs.get_initial_alias()
-        rhs.bump_prefix(self, exclude={
-            alias for alias, join in rhs.alias_map.items()
-            if join.table_alias in [join.table_name, initial_alias]
-        })
-        rhs.unref_alias(initial_alias)
+        rhs.bump_prefix(self, exclude={self.get_initial_alias()})
 
         # Work out how to relabel the rhs aliases, if necessary.
         change_map = {}
@@ -598,9 +593,6 @@ class Query(BaseExpression):
         # the AND case. The results will be correct but this creates too many
         # joins. This is something that could be fixed later on.
         reuse = set() if conjunction else set(self.alias_map)
-        # Base table must be present in the query - this is the same
-        # table on both sides.
-        self.get_initial_alias()
         joinpromoter = JoinPromoter(connector, 2, False)
         joinpromoter.add_votes(
             j for j in self.alias_map if self.alias_map[j].join_type == INNER)

```

Reply:
If we only exclude the initial alias then tables that previously had no alias will get an alias for the `rhs`. I thought this might have some side-effects (while reusing joins for example), and wanted to exclude tables that have no alias. If we are sure that this won't be a problem, then I am refactoring the related part according to your comment.